File: mediatek/frameworks/opt/ngin3d/java/com/mediatek/ngin3d/Layer.java
/* Copyright Statement:
 *
 * This software/firmware and related documentation ("MediaTek Software") are
 * protected under relevant copyright laws. The information contained herein is
 * confidential and proprietary to MediaTek Inc. and/or its licensors. Without
 * the prior written permission of MediaTek inc. and/or its licensors, any
 * reproduction, modification, use or disclosure of MediaTek Software, and
 * information contained herein, in whole or in part, shall be strictly
 * prohibited.
 *
 * MediaTek Inc. (C) 2013. All rights reserved.
 *
 * BY OPENING THIS FILE, RECEIVER HEREBY UNEQUIVOCALLY ACKNOWLEDGES AND AGREES
 * THAT THE SOFTWARE/FIRMWARE AND ITS DOCUMENTATIONS ("MEDIATEK SOFTWARE")
 * RECEIVED FROM MEDIATEK AND/OR ITS REPRESENTATIVES ARE PROVIDED TO RECEIVER
 * ON AN "AS-IS" BASIS ONLY. MEDIATEK EXPRESSLY DISCLAIMS ANY AND ALL
 * WARRANTIES, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR
 * NONINFRINGEMENT. NEITHER DOES MEDIATEK PROVIDE ANY WARRANTY WHATSOEVER WITH
 * RESPECT TO THE SOFTWARE OF ANY THIRD PARTY WHICH MAY BE USED BY,
 * INCORPORATED IN, OR SUPPLIED WITH THE MEDIATEK SOFTWARE, AND RECEIVER AGREES
 * TO LOOK ONLY TO SUCH THIRD PARTY FOR ANY WARRANTY CLAIM RELATING THERETO.
 * RECEIVER EXPRESSLY ACKNOWLEDGES THAT IT IS RECEIVER'S SOLE RESPONSIBILITY TO
 * OBTAIN FROM ANY THIRD PARTY ALL PROPER LICENSES CONTAINED IN MEDIATEK
 * SOFTWARE. MEDIATEK SHALL ALSO NOT BE RESPONSIBLE FOR ANY MEDIATEK SOFTWARE
 * RELEASES MADE TO RECEIVER'S SPECIFICATION OR TO CONFORM TO A PARTICULAR
 * STANDARD OR OPEN FORUM. RECEIVER'S SOLE AND EXCLUSIVE REMEDY AND MEDIATEK'S
 * ENTIRE AND CUMULATIVE LIABILITY WITH RESPECT TO THE MEDIATEK SOFTWARE
 * RELEASED HEREUNDER WILL BE, AT MEDIATEK'S OPTION, TO REVISE OR REPLACE THE
 * MEDIATEK SOFTWARE AT ISSUE, OR REFUND ANY SOFTWARE LICENSE FEES OR SERVICE
 * CHARGE PAID BY RECEIVER TO MEDIATEK FOR SUCH MEDIATEK SOFTWARE AT ISSUE.
 *
 * The following software/firmware and/or related documentation ("MediaTek
 * Software") have been modified by MediaTek Inc. All revisions are subject to
 * any receiver's applicable license agreements with MediaTek Inc.
 */

package com.mediatek.ngin3d;

import com.mediatek.ngin3d.presentation.Presentation;
import com.mediatek.ngin3d.presentation.PresentationEngine;
import com.mediatek.ngin3d.presentation.PresentationHitTestResult;
import com.mediatek.ngin3d.presentation.RenderLayer;

/**
 * Layer is a special Container that displays its Actors on the screen.
 * Each Layer object has its own camera settings. Layer objects are added to
 * the Stage object. They are drawn in the order that they are added to the
 * Stage.
 *
 * \ingroup ngin3dClasses
 */
public class Layer extends Container {
    protected static final String TAG = "Layer";
    /*
     * Projection modes
     */
    /** Orthographic projection mode. */
    public static final int ORTHOGRAPHIC = 0;
    /** Perspective projection mode. */
    public static final int PERSPECTIVE = 1;

    private RenderLayer mRenderLayer;
    private Point mPosition;

    /**
     * @hide
     */
    @Override
    protected Presentation createPresentation(PresentationEngine engine) {
        Presentation p = super.createPresentation(engine);
        p.initialize(this);
        if (mRenderLayer == null) {
            mRenderLayer = engine.createRenderLayer(p);
        }
        return p;
    }

    public RenderLayer getRenderLayer() {
        return mRenderLayer;
    }

    /**
     * @hide
     */
    static final Property<Point> PROP_CAMERA_POS = new Property<Point>(
            "camera", null);
    /**
     * @hide
     */
    static final Property<Rotation> PROP_CAMERA_ROT = new Property<Rotation>(
            "camera_rotation", null);
    /**
     * @hide
     */
    static final Property<Float> PROP_CAMERA_FOV = new Property<Float>(
            "camera_fov", null);
    /**
     * @hide
     */
    static final Property<Float> PROP_CAMERA_NEAR = new Property<Float>(
            "camera_near", null);
    /**
     * @hide
     */
    static final Property<Float> PROP_CAMERA_FAR = new Property<Float>(
            "camera_far", null);
    /**
     * @hide
     */
    static final Property<Float> PROP_CAMERA_WIDTH = new Property<Float>(
            "camera_width", null);
    /**
     * @hide
     */
    static final Property<Integer> PROP_PROJECTION_MODE = new Property<Integer>(
            "proj_mode", null);
    /**
     * @hide
     */
    static final Property<Boolean> PROP_CLEAR_DEPTH = new Property<Boolean>(
            "clear_depth", null);
    /**
     * @hide
     */
    static final Property<String> PROP_NAMED_CAMERA = new Property<String>(
            "named_camera", null);


    protected boolean applyValue(Property property, Object value) {
        if (super.applyValue(property, value)) {
            return true;
        }

        if (property.sameInstance(PROP_CAMERA_POS)) {
            if (value != null) {
                Point position = (Point) value;
                mRenderLayer.setCameraPosition(position);
            }
            return true;
        } else if (property.sameInstance(PROP_CAMERA_ROT)) {
            if (value != null) {
                Rotation rotation = (Rotation) value;
                mRenderLayer.setCameraRotation(rotation);
            }
            return true;
        } else if (property.sameInstance(PROP_CAMERA_FOV)) {
            if (value != null) {
                Float fov = (Float) value;
                mRenderLayer.setCameraFov(fov);
            }
            return true;
        } else if (property.sameInstance(PROP_CAMERA_NEAR)) {
            if (value != null) {
                Float near = (Float) value;
                mRenderLayer.setCameraNear(near);
            }
            return true;
        } else if (property.sameInstance(PROP_CAMERA_FAR)) {
            if (value != null) {
                Float far = (Float) value;
                mRenderLayer.setCameraFar(far);
            }
            return true;
        } else if (property.sameInstance(PROP_CAMERA_WIDTH)) {
            if (value != null) {
                Float far = (Float) value;
                mRenderLayer.setCameraWidth(far);
            }
            return true;
        } else if (property.sameInstance(PROP_PROJECTION_MODE)) {
            if (value != null) {
                Integer mode = (Integer) value;
                mRenderLayer.setProjectionMode(mode);
            }
            return true;
        } else if (property.sameInstance(PROP_NAMED_CAMERA)) {
            if (value != null) {
                String name = (String) value;
                mRenderLayer.useNamedCamera(name);
            }
            return true;
        } else if (property.sameInstance(PROP_CLEAR_DEPTH)) {
            if (value != null) {
                mRenderLayer.setDepthClear((Boolean)value);
            }
            return true;
        }

        return false;
    }

    /**
     * Sets the camera position.
     * The camera will be set to the given position. The orientation is
     * unaffected.
     *
     * @param position Camera position
     */
    public void setCameraPosition(Point position) {
        mPosition = position;
        setValueInTransaction(PROP_CAMERA_POS, position);
    }

    /**
     * Sets the camera orientation.
     * The camera will oriented to the given rotation.
     *
     * @param rotation Camera orientation
     */
    public void setCameraRotation(Rotation rotation) {
        setValueInTransaction(PROP_CAMERA_ROT, rotation);
    }

    /**
     * Sets the camera position and orientation.
     * The camera will be set to the given position and oriented so that
     * it is pointing towards the "lookAt" point. If you subsequently change
     * the camera position it will not re-orient itself to look at the given
     * point. The subsequent camera rotation will have its "up" vector as
     * close to (0,1,0) as possible.
     *
     * @param lookAt    Camera focus point
     */
    public void setCameraLookAt(Point lookAt) {
        setCameraLookAt(lookAt, new Point(0, 1, 0));
    }

    /**
     * Sets the camera position and orientation.
     * The camera will be set to the given position and oriented so that
     * it is pointing towards the "lookAt" point. If you subsequently change
     * the camera position it will not re-orient itself to look at the given
     * point.
     *
     * @param lookAt    Camera focus point
     * @param up        Vector to keep up (usually (0,1,0) or (0,-1,0))
     */
    public void setCameraLookAt(Point lookAt, Point up) {
        Point to = new Point(lookAt.x - mPosition.x,
                lookAt.y - mPosition.y, lookAt.z - mPosition.z);
        Rotation rotation = Rotation.pointAt(new Point(0, 0, -1),
                to, new Point(0, 1, 0), new Point(up.x, up.y, up.z));
        setValueInTransaction(PROP_CAMERA_ROT, rotation);
    }

    /**
     * Set camera field of view (FOV) in degrees.
     * The field of view for the smaller screen dimension is specified (e.g. if
     * the screen is taller than it is wide, the horizontal FOV is specified).
     *
     * This parameter is only used by the PERSPECTIVE projection. In the 'UI'
     * projections the FOV is derived from the camera Z position and the screen
     * width (pixels) which are considered to be in the same coordinate space.
     *
     * @param fov Camera field of view in degrees
     */
    public void setCameraFov(float fov) {
        setValueInTransaction(PROP_CAMERA_FOV, fov);
    }

    /**
     * Set camera near clipping plane.
     *
     * @param near Camera near clipping plane
     */
    public void setCameraNear(float near) {
        setValueInTransaction(PROP_CAMERA_NEAR, near);
    }

    /**
     * Set camera width.
     * Sets the width of the viewing frustum in world-units when using an
     * ORTHOGRAPHIC projection.
     *
     * When using an PERSPECTIVE projection, this parameter has no visible
     * effect.
     *
     * @param width Width of viewing frustum
     */
    public void setCameraWidth(float width) {
        setValueInTransaction(PROP_CAMERA_WIDTH, width);
    }

    /**
     * Set camera far clipping plane.
     *
     * @param far Camera far clipping plane
     */
    public void setCameraFar(float far) {
        setValueInTransaction(PROP_CAMERA_FAR, far);
    }

    /**
     * Set projection mode
     *
     * @param mode Camera projection mode ORTHOGRAPHIC or PERSPECTIVE
     */
    public void setProjectionMode(int mode) {
        setValueInTransaction(PROP_PROJECTION_MODE, mode);
    }

    /**
     * Sets depth buffer clearing behaviour.
     * By default the depth buffer is cleared before each layer is rendered.
     * Call this method with false to prevent the z-buffer being cleared before
     * the layer is rendered.
     *
     * @param clear true if the depth buffer should be cleared
     */
    public void setDepthClear(boolean clear) {
        setValueInTransaction(PROP_CLEAR_DEPTH, clear);
    }

    /**
     * Use a camera from the scene
     * Use a camera created in a DCC tool and exported to a glo file. You should
     * pass in the name you gave the camera in 3ds Max or Blender. Pass an empty
     * string to revert to using the Layer's camera parameters.
     *
     * @param name Camera in scene to use
     */
    public void useNamedCamera(String name) {
        setValueInTransaction(PROP_NAMED_CAMERA, name);
    }

    /**
     * Use UI perspective projection
     * The camera will be positioned at the given (zPos) distance from the
     * X/Y plane and the field of view set such that the camera covers an
     * area of the plane so that x=0 corresponds to the left of the screen,
     * x=width corresponds to the right of the screen, y=0 corresponds to
     * the top of the screen and y=height corresponds to the bottom of the
     * screen.
     *
     * @param width     width of screen
     * @param height    height of screen
     * @param zPos      distance of camera from X/Y plane
     */
    public void setUiPerspective(float width, float height, float zPos) {
        float smallerDim = Math.min(width, height);
        float zDistance = Math.abs(zPos);

        float fov = (float) Math.toDegrees(
            Math.atan((smallerDim / 2) / zDistance) * 2);
        setCameraFov(fov);

        float centX = width / 2;
        float centY = height / 2;

        Point cameraPosition = new Point(centX, centY, zPos);

        setCameraPosition(cameraPosition);

        float zDir = (zPos < 0) ? 1 : -1;
        Rotation rotation = Rotation.pointAt(new Point(0, 0, -1),
                new Point(0, 0, zDir),
                new Point(0, 1, 0),
                new Point(0, -1, 0));
        setCameraRotation(rotation);

        // Set near and far planes such that the z=0 plane is half-way between
        // them and the ratio near/far = 1/100
        setCameraNear(2 * zDistance / 101);
        setCameraFar(200 * zDistance / 101);
    }

    /**
     * Used by hitTestFull()
     * Overridden by this class to provide a RenderLayer object (and perform
     * the hit test with the associated camera).
     *
     * @param screenPoint Point on the screen to test
     * @return Details about the hit test and its result
     */
    @Override
    protected PresentationHitTestResult doHitTest(Point screenPoint) {
        return mPresentation.hitTest(screenPoint, mRenderLayer);
    }

}

File: mediatek/frameworks-ext/native/services/surfaceflinger/Layer.cpp
/* Copyright Statement:
 *
 * This software/firmware and related documentation ("MediaTek Software") are
 * protected under relevant copyright laws. The information contained herein is
 * confidential and proprietary to MediaTek Inc. and/or its licensors. Without
 * the prior written permission of MediaTek inc. and/or its licensors, any
 * reproduction, modification, use or disclosure of MediaTek Software, and
 * information contained herein, in whole or in part, shall be strictly
 * prohibited.
 * 
 * MediaTek Inc. (C) 2010. All rights reserved.
 * 
 * BY OPENING THIS FILE, RECEIVER HEREBY UNEQUIVOCALLY ACKNOWLEDGES AND AGREES
 * THAT THE SOFTWARE/FIRMWARE AND ITS DOCUMENTATIONS ("MEDIATEK SOFTWARE")
 * RECEIVED FROM MEDIATEK AND/OR ITS REPRESENTATIVES ARE PROVIDED TO RECEIVER
 * ON AN "AS-IS" BASIS ONLY. MEDIATEK EXPRESSLY DISCLAIMS ANY AND ALL
 * WARRANTIES, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR
 * NONINFRINGEMENT. NEITHER DOES MEDIATEK PROVIDE ANY WARRANTY WHATSOEVER WITH
 * RESPECT TO THE SOFTWARE OF ANY THIRD PARTY WHICH MAY BE USED BY,
 * INCORPORATED IN, OR SUPPLIED WITH THE MEDIATEK SOFTWARE, AND RECEIVER AGREES
 * TO LOOK ONLY TO SUCH THIRD PARTY FOR ANY WARRANTY CLAIM RELATING THERETO.
 * RECEIVER EXPRESSLY ACKNOWLEDGES THAT IT IS RECEIVER'S SOLE RESPONSIBILITY TO
 * OBTAIN FROM ANY THIRD PARTY ALL PROPER LICENSES CONTAINED IN MEDIATEK
 * SOFTWARE. MEDIATEK SHALL ALSO NOT BE RESPONSIBLE FOR ANY MEDIATEK SOFTWARE
 * RELEASES MADE TO RECEIVER'S SPECIFICATION OR TO CONFORM TO A PARTICULAR
 * STANDARD OR OPEN FORUM. RECEIVER'S SOLE AND EXCLUSIVE REMEDY AND MEDIATEK'S
 * ENTIRE AND CUMULATIVE LIABILITY WITH RESPECT TO THE MEDIATEK SOFTWARE
 * RELEASED HEREUNDER WILL BE, AT MEDIATEK'S OPTION, TO REVISE OR REPLACE THE
 * MEDIATEK SOFTWARE AT ISSUE, OR REFUND ANY SOFTWARE LICENSE FEES OR SERVICE
 * CHARGE PAID BY RECEIVER TO MEDIATEK FOR SUCH MEDIATEK SOFTWARE AT ISSUE.
 *
 * The following software/firmware and/or related documentation ("MediaTek
 * Software") have been modified by MediaTek Inc. All revisions are subject to
 * any receiver's applicable license agreements with MediaTek Inc.
 */

#include <cutils/xlog.h>
#include <cutils/properties.h>

#include <SkImageEncoder.h>
#include <SkBitmap.h>

#include <gui/ISurfaceComposer.h>
#include <gui/BufferQueue.h>

#include "SurfaceFlinger.h"
#include "Layer.h"

namespace android {

// dump current using buffer in Layer
void Layer::dumpActiveBuffer() const {
    XLOGV("[dumpActiveBuffer] + id=%d", getIdentity());

    if (mActiveBuffer != NULL) {
        char     value[PROPERTY_VALUE_MAX];
        bool     raw;
        uint32_t identity;

        property_get("debug.sf.layerdump.raw", value, "0");
        raw = (0 != atoi(value));
        identity = getIdentity();

        char             fname[128];
        void*            ptr;
        uint32_t         size;
        SkBitmap         b;
        SkBitmap::Config c;

        size = 0;
        c = SkBitmap::kNo_Config;
        switch (mActiveBuffer->format) {
            case PIXEL_FORMAT_RGBA_8888:
            case PIXEL_FORMAT_RGBX_8888:
                if (false == raw) {
                    c = SkBitmap::kARGB_8888_Config;
                    sprintf(fname, "/data/SF_dump/%d.png", identity);
                } else {
                    size = mActiveBuffer->stride * mActiveBuffer->height * 4;
                    sprintf(fname, "/data/SF_dump/%d.RGBA", identity);
                }
                break;
            case PIXEL_FORMAT_BGRA_8888:
            case 0x1ff:                     // tricky format for SGX_COLOR_FORMAT_BGRX_8888 in fact
                if (false == raw) {
                    c = SkBitmap::kARGB_8888_Config;
                    sprintf(fname, "/data/SF_dump/%d(RBswapped).png", identity);
                } else {
                    size = mActiveBuffer->stride * mActiveBuffer->height * 4;
                    sprintf(fname, "/data/SF_dump/%d.BGRA", identity);
                }
                break;
            case PIXEL_FORMAT_RGB_565:
                if (false == raw) {
                    c = SkBitmap::kRGB_565_Config;
                    sprintf(fname, "/data/SF_dump/%d.png", identity);
                } else {
                    size = mActiveBuffer->stride * mActiveBuffer->height * 2;
                    sprintf(fname, "/data/SF_dump/%d.RGB565", identity);
                }
                break;
            case HAL_PIXEL_FORMAT_I420:
                size = mActiveBuffer->stride * mActiveBuffer->height * 3 / 2;
                sprintf(fname, "/data/SF_dump/%d.i420", identity);
                break;
            case HAL_PIXEL_FORMAT_NV12_BLK:
                size = mActiveBuffer->stride * mActiveBuffer->height * 3 / 2;
                sprintf(fname, "/data/SF_dump/%d.nv12_blk", identity);
                break;
            case HAL_PIXEL_FORMAT_NV12_BLK_FCM:
                size = mActiveBuffer->stride * mActiveBuffer->height * 3 / 2;
                sprintf(fname, "/data/SF_dump/%d.nv12_blk_fcm", identity);
                break;
            case HAL_PIXEL_FORMAT_YV12:
                // android YV12 need align to 16 for UV plane
                size = (mActiveBuffer->stride * mActiveBuffer->height) + 
                       ((mActiveBuffer->stride / 2 + 0xf) & (~0xf)) * (mActiveBuffer->height / 2) * 2;
                sprintf(fname, "/data/SF_dump/%d.yv12", identity);
                break;
            default:
                XLOGE("[%s] cannot dump format:%d for identity:%d",
                      __func__, mActiveBuffer->format, identity);
                return;
        }

        mActiveBuffer->lock(GraphicBuffer::USAGE_SW_READ_OFTEN, &ptr);
        {
            XLOGI("[%s] %s", __func__, getName().string());
            XLOGI("    %s (config:%d, stride:%d, height:%d, ptr:%p)",
                fname, c, mActiveBuffer->stride, mActiveBuffer->height, ptr);

            if (SkBitmap::kNo_Config != c) {
                b.setConfig(c, mActiveBuffer->stride, mActiveBuffer->height);
                b.setPixels(ptr);
                SkImageEncoder::EncodeFile(fname, b, SkImageEncoder::kPNG_Type,
                                           SkImageEncoder::kDefaultQuality);
            } else {
                FILE *f = fopen(fname, "wb");
                fwrite(ptr, size, 1, f);
                fclose(f);
            }
        }
        mActiveBuffer->unlock();

        dumpContinuousBuffer();
    }
    XLOGV("[dumpActiveBuffer] - id=%d", getIdentity());
}

void Layer::setContBufsDumpById(int identity) {
    XLOGV("setContBufsDumpById, id=%d", identity);
    mContBufsDumpById = identity;
}

void Layer::activeBufferBackup() {
    if (mActiveBuffer == NULL) {
        XLOGW("[Layer::activeBufferBackup] mActiveBuffer=%p not initialized", mActiveBuffer.get());
        return;
    }
    
    if (mContBufsDumpById == (int)getIdentity() || mContBufsDumpById == -1) {
        XLOGV("[Layer::activeBufferBackup] +, req=%d, id=%d", mContBufsDumpById, getIdentity());
        // check bpp
        float bpp = 0.0f;
        uint32_t width  = mActiveBuffer->width;
        uint32_t height = mActiveBuffer->height;
        uint32_t format = mActiveBuffer->format;
        uint32_t usage  = mActiveBuffer->usage;
        uint32_t stride = mActiveBuffer->stride;
        status_t err;
        
        switch (mActiveBuffer->format) {
            case PIXEL_FORMAT_RGBA_8888:
            case PIXEL_FORMAT_BGRA_8888:
            case PIXEL_FORMAT_RGBX_8888:
            case 0x1ff:
                // tricky format for SGX_COLOR_FORMAT_BGRX_8888 in fact
                bpp = 4.0;
                break;
            case PIXEL_FORMAT_RGB_565:
                bpp = 2.0;
                break;
            case HAL_PIXEL_FORMAT_I420:
                bpp = 1.5;
                break;
            case HAL_PIXEL_FORMAT_YV12:
                bpp = 1.5;
                break;
            default:
                XLOGE("[%s] cannot dump format:%d for identity:%d", __func__, mActiveBuffer->format, getIdentity());
                break;
        }

#define MAX_DEFAULT_BUFFERS 10
        // initialize backup buffer max size
        char value[PROPERTY_VALUE_MAX];

        property_get("debug.sf.contbufsmax", value, "0");
        uint32_t max = atoi(value);
        if (max <= 0)
            max = MAX_DEFAULT_BUFFERS;

        if (mBackupBufsMax != max) {
            mBackupBufsMax = max;
            XLOGI("==>  ring buffer max size, max = %d", max);

            mBackBufs.clear();
            mBackupBufsIndex = 0;
        }

        // create new GraphicBuffer
        if (mBackBufs.size() < mBackupBufsMax) {
            sp<GraphicBuffer> buf;
            buf = new GraphicBuffer(width, height, format, usage);
            mBackBufs.push(buf);

            XLOGI("[id=%d] new buffer for cont. dump, size = %d", getIdentity(), mBackBufs.size());
        }
        
        // detect geometry changed
        if (mBackBufs[mBackupBufsIndex]->width != mActiveBuffer->width || 
            mBackBufs[mBackupBufsIndex]->height != mActiveBuffer->height ||
            mBackBufs[mBackupBufsIndex]->format != mActiveBuffer->format) {
            XLOGI("[id=%d] geometry changed, backup=(%d, %d, %d) ==> active=(%d, %d, %d)",
                getIdentity(),
                mBackBufs[mBackupBufsIndex]->width,
                mBackBufs[mBackupBufsIndex]->height,
                mBackBufs[mBackupBufsIndex]->format,
                mActiveBuffer->width,
                mActiveBuffer->height,
                mActiveBuffer->format);

            sp<GraphicBuffer> buf;
            buf = new GraphicBuffer(width, height, format, usage);
            mBackBufs.replaceAt(buf, mBackupBufsIndex);
        }

        if (mActiveBuffer.get() == NULL || mBackBufs[mBackupBufsIndex] == NULL) {
            XLOGW("[Layer::activeBufferBackup] backup fail, mActiveBuffer=%p, mBackBufs[%d]=%p",
                mActiveBuffer.get(), mBackupBufsIndex, mBackBufs[mBackupBufsIndex].get());
            return;
        }
        
        // backup
        nsecs_t now = systemTime();        

        void* src;
        void* dst;
        err = mActiveBuffer->lock(GraphicBuffer::USAGE_SW_READ_OFTEN, &src);
        if (err != NO_ERROR) {
            XLOGW("[Layer::activeBufferBackup] lock active buffer failed");
            return;
        }

        XLOGV("[Layer::activeBufferBackup] lock +, req=%d, id=%d", mContBufsDumpById, getIdentity());
        err = mBackBufs[mBackupBufsIndex]->lock(GraphicBuffer::USAGE_SW_READ_OFTEN | GraphicBuffer::USAGE_SW_WRITE_OFTEN, &dst);
        if (err != NO_ERROR) {
            XLOGW("[Layer::activeBufferBackup] lock backup buffer failed");
            return;
        }

        backupProcess(dst, src, stride*height*bpp);

        mBackBufs[mBackupBufsIndex]->unlock();
        mActiveBuffer->unlock();

        if (mContBufsDumpById == -1) {
            XLOGI("[Layer::activeBufferBackup] req=%d, id=%d, buf=%d, time=%lld", 
                mContBufsDumpById, getIdentity(), mBackupBufsIndex, ns2ms(systemTime() - now));
        } else {
            XLOGI("[Layer::activeBufferBackup] id=%d, buf=%d, time=%lld", 
                getIdentity(), mBackupBufsIndex, ns2ms(systemTime() - now));
        }
        mBackupBufsIndex ++;
        if (mBackupBufsIndex >= mBackupBufsMax)
            mBackupBufsIndex = 0;
    }
}

void Layer::backupProcess(void* dst, void* src, size_t size) const {
    XLOGV("[Layer::backupProcess] +, req=%d, id=%d", mContBufsDumpById, getIdentity());

    // backup 
    memcpy(dst, src, size);

    XLOGV("[Layer::backupProcess] -, req=%d, id=%d", mContBufsDumpById, getIdentity());
}

void Layer::dumpContinuousBuffer() const {
    char tmp[PROPERTY_VALUE_MAX];
    int  identity = getIdentity();

    if (mContBufsDumpById <= 0 && mContBufsDumpById != -1)
        return;
        
    if (mBackupBufsMax <= 0) {
        XLOGW("[Layer::dumpContinuousBuffer] mBackupBufsMax not updated");
        return;
    }

    XLOGD("[Layer::dumpContinuousBuffer] +, req=%d, id=%d, size=%d",
        mContBufsDumpById, getIdentity(), mBackBufs.size());

    if (mContBufsDumpById == (int)getIdentity() || mContBufsDumpById == -1) {
        int start = (mBackupBufsIndex + mBackupBufsMax - 1) % mBackupBufsMax;
        for (uint32_t i = 0; i < mBackupBufsMax; i++) {
            if (i >= mBackBufs.size()) {
                XLOGW("[Layer::dumpContinuousBuffer] overflow i=%d, max=%d", i, mBackBufs.size());
                return;
            }

            int index = (start + mBackupBufsMax - i) % mBackupBufsMax;
            XLOGD("[Layer::dumpContinuousBuffer] i=%d, index=%d", mBackupBufsMax - i, index);
            sp<GraphicBuffer> buf = mBackBufs[index];
            dumpGraphicBuffer(buf, mBackupBufsMax - i);
        }
    }

    XLOGD("[Layer::dumpContinuousBuffer] -");
}

void Layer::dumpGraphicBuffer(sp<GraphicBuffer> buf, int index) const {
    char             fname[128];
    void*            ptr;
    SkBitmap         b;
    SkBitmap::Config c;
    int              identity = getIdentity();
    float            bpp;
    
    
    c = SkBitmap::kNo_Config;
    switch (buf->format) {
        case PIXEL_FORMAT_RGBA_8888:
        case PIXEL_FORMAT_BGRA_8888:
        case PIXEL_FORMAT_RGBX_8888:
        case 0x1ff:                     // tricky format for SGX_COLOR_FORMAT_BGRX_8888 in fact
            c = SkBitmap::kARGB_8888_Config;
            sprintf(fname, "/data/SF_dump/%d_%03d.png", identity, index);
            break;
        case PIXEL_FORMAT_RGB_565:
            c = SkBitmap::kRGB_565_Config;
            sprintf(fname, "/data/SF_dump/%d_%03d.png", identity, index);
            break;
        case HAL_PIXEL_FORMAT_I420:
            bpp = 1.5;
            sprintf(fname, "/data/SF_dump/%d_%03d.i420", identity, index);
            break;
        case HAL_PIXEL_FORMAT_YV12:
            bpp = 1.5;
            sprintf(fname, "/data/SF_dump/%d_%03d.yv12", identity, index);
            break;
        default:
            XLOGE("[%s] cannot dump format:%d for identity:%d", __func__, buf->format, identity);
            return;
    }

    buf->lock(GraphicBuffer::USAGE_SW_READ_OFTEN, &ptr);
    {
        XLOGI("[Layer::dumpGraphicBuffer] %s", getName().string());
        XLOGI("    %s (config:%d, stride:%d, height:%d, ptr:%p)",
            fname, c, buf->stride, buf->height, ptr);

        if (SkBitmap::kNo_Config != c) {
            b.setConfig(c, buf->stride, buf->height);
            b.setPixels(ptr);
            SkImageEncoder::EncodeFile(
                fname, b, SkImageEncoder::kPNG_Type, SkImageEncoder::kDefaultQuality);
        } else {
            uint32_t size = buf->stride * buf->height * bpp;
            FILE *f = fopen(fname, "wb");
            fwrite(ptr, size, 1, f);
            fclose(f);
        }
    }
    buf->unlock();
}

};
